Possible race condition in evc.deploy() when run through execute() in DEPLOY_EVCS_INTERVAL

Hi,

The problem is that an EVC can have more than two OpenFlow rules installed at the switch's flow table, for instance:
```
>               assert len(re.findall(evc['id'], flows_s1, flags=re.IGNORECASE)) == 2
E               assert 3 == 2
E                +  where 3 = len(['7e4621078b7844', '7e4621078b7844', '7e4621078b7844'])
E                +    where ['7e4621078b7844', '7e4621078b7844', '7e4621078b7844'] = <function findall at 0x7faa83384f50>('7e4621078b7844', ' cookie=0xaa2e5bdbedaa494d, duration=11.457s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=400 actions=...n=68.643s, table=0, n_packets=44, n_bytes=1848, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n', flags=2)
E                +      where <function findall at 0x7faa83384f50> = re.findall
E                +      and   2 = re.IGNORECASE
...
E                  cookie=0xaa7e4621078b7844, duration=10.323s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth3",dl_vlan=3761 actions=strip_vlan,mod_vlan_vid:406,output:"s1-eth1"
E                  cookie=0xaa7e4621078b7844, duration=10.317s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth3",dl_vlan=1066 actions=strip_vlan,mod_vlan_vid:406,output:"s1-eth1"
E                  cookie=0xaa7e4621078b7844, duration=0.478s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=406 actions=strip_vlan,mod_vlan_vid:1066,output:"s1-eth3"
E                  cookie=0x0, duration=68.643s, table=0, n_packets=44, n_bytes=1848, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
```

As we can see on the output above, the EVC `0xaa7e4621078b7844 ` has three flows: two of them using the service vlan tag 1066 and anther one with service vlan 3761. In fact, I suppose the correct service vlan would be 3761, however it is possible that the consistency check has overwritten the UNI ingress flow (since the match will be the same).

The output above is part of the following end-to-end test:
```
+    def test_070_repeat_creation_and_removal_of_ten_circuits_ten_times(self):
+        for x in range(0, 10):
+            evcs = {}
+            for i in range(400, 410):
+                payload = {
+                    "name": "evc_%s" % i,
+                    "enabled": True,
+                    "dynamic_backup_path": True,
+                    "uni_a": {
+                        "interface_id": "00:00:00:00:00:00:00:01:1",
+                        "tag": {"tag_type": 1, "value": i}
+                    },
+                    "uni_z": {
+                        "interface_id": "00:00:00:00:00:00:00:02:1",
+                        "tag": {"tag_type": 1, "value": i}
+                    }
+                }
+                api_url = KYTOS_API + '/mef_eline/v2/evc/'
+                response = requests.post(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
+                assert response.status_code == 201
+                data = response.json()
+                assert 'circuit_id' in data
+                evcs[i] = data['circuit_id']
+
+            time.sleep(10)
+
+            # make sure the evcs are active and the flows were created
+            s1, s2 = self.net.net.get('s1', 's2')
+            flows_s1 = s1.dpctl('dump-flows')
+            flows_s2 = s2.dpctl('dump-flows')
+            for vid in evcs:
+                evc_id = evcs[vid]
+                api_url = KYTOS_API + '/mef_eline/v2/evc/' + evc_id
+                response = requests.get(api_url)
+                assert response.status_code == 200
+                evc = response.json()
+                #should be active
+                assert evc["active"] is True
+                # search for the vlan id
+                assert "dl_vlan=%s" % vid in flows_s1
+                assert "dl_vlan=%s" % vid in flows_s2
+                # search for the cookie, should have two flows
+                assert len(re.findall(evc['id'], flows_s1, flags=re.IGNORECASE)) == 2, "round=%d - should have 2 flows but had: \n%
s" % (x, flows_s1)
+                assert len(re.findall(evc['id'], flows_s2, flags=re.IGNORECASE)) == 2, "round=%d - should have 2 flows but had: \n%
s" % (x, flows_s2)
+
+            # Delete the circuits
+            for vid in evcs:
+                evc_id = evcs[vid]
+                api_url = KYTOS_API + '/mef_eline/v2/evc/' + evc_id
+                response = requests.delete(api_url)
+                assert response.status_code == 200
+
+            time.sleep(10)
+
+            # make sure the circuits were deleted
+            api_url = KYTOS_API + '/mef_eline/v2/evc/'
+            response = requests.get(api_url)
+            assert response.status_code == 200
+            assert response.json() == {}
+            flows_s1 = s1.dpctl('dump-flows')
+            flows_s2 = s2.dpctl('dump-flows')
+            assert len(flows_s1.split('\r\n ')) == 1, "round=%d - should have only 1 flow but had: \n%s" % (x, flows_s1)
+            assert len(flows_s2.split('\r\n ')) == 1, "round=%d - should have only 1 flow but had: \n%s" % (x, flows_s2)
```

Decreasing the setting DEPLOY_EVCS_INTERVAL only makes the result more frequent.

It looks like this is another race condition. Mef_eline has its own "consistency check routine" which is run through execute() call based on the setting DEPLOY_EVCS_INTERVAL. This routine basically searches for undeployed EVCs (i.e., enabled but not active) and tries deploy them. The problem arrives when multiple EVCs are created in a repeated way such that they may be requested at the same time that the consistency is running.

--- a/main.py
+++ b/main.py
@@ -58,7 +58,7 @@ class Main(KytosNApp):
         """Execute once when the napp is running."""
         for circuit in tuple(self.circuits.values()):
             if circuit.is_enabled() and not circuit.is_active():
-                with evc.lock:
+                with circuit.lock:
                     circuit.deploy()
 
     def shutdown(self):

--- a/main.py
+++ b/main.py
@@ -58,7 +58,8 @@ class Main(KytosNApp):
         """Execute once when the napp is running."""
         for circuit in tuple(self.circuits.values()):
             if circuit.is_enabled() and not circuit.is_active():
-                circuit.deploy()
+                with evc.lock:
+                    circuit.deploy()
 
     def shutdown(self):
         """Execute when your napp is unloaded.
@@ -162,7 +163,8 @@ class Main(KytosNApp):
 
         # Circuit has no schedule, deploy now
         if not evc.circuit_scheduler:
-            evc.deploy()
+            with evc.lock:
+                evc.deploy()
 
         # Notify users
         event = KytosEvent(name='kytos.mef_eline.created',
@@ -216,11 +218,13 @@ class Main(KytosNApp):
             if enable is False:  # disable if active
                 evc.remove()
             elif path is not None:  # redeploy if active
-                evc.remove()
-                evc.deploy()
+                with evc.lock:
+                    evc.remove()
+                    evc.deploy()
         else:
             if enable is True:  # enable if inactive
-                evc.deploy()
+                with evc.lock:
+                    evc.deploy()
         result = {evc.id: evc.as_dict()}
         status = 200
 

--- a/main.py
+++ b/main.py
@@ -216,7 +216,8 @@ class Main(KytosNApp):
 
         if evc.is_active():
             if enable is False:  # disable if active
-                evc.remove()
+                with evc.lock:
+                    evc.remove()
             elif path is not None:  # redeploy if active
                 with evc.lock:
                     evc.remove()

